File: app/validate.py
# -*- coding: utf-8 -*-
"""external inputs validation"""

import os
import re

id_check = re.compile('([0-9a-f]+)')
zip_check = re.compile('([0-9a-zA-Z_.-]+.zip)')
fb2_check = re.compile('([ 0-9a-zA-ZА-Яа-я_,.:!-]+.fb2)')  # may be incomplete
genre_check = re.compile('([0-9a-z_]+)')


def unurl(string: str):
    """very simple url to string"""
    translate = {
        '%22': '"',
        '%27': "'",
        '%2E': ".",
        '%2F': '/'
    }
    ret = string
    if ret is not None:
        for k, v in translate.items():  # pylint: disable=C0103
            ret = ret.replace(k, v)
    return ret


def safe_path(fspath):
    """create safe relative path from input"""
    if fspath is None:
        return None
    return os.path.relpath(os.path.normpath(os.path.join("/", fspath)), "/")


def validate_prefix(string: str):
    """very simple prefix validation in .../sequenceindes and .../authorsindex"""
    if string is None:
        return None
    ret = safe_path(string)
    if len(ret) > 10 or len(ret) < 1:
        return None
    return ret


def validate_id(string: str):
    """author/book/sequence id validation"""
    ret = string
    if id_check.match(string):
        return ret
    return None


def validate_zip(string: str):
    """zip filename validation"""
    ret = string
    if zip_check.match(string):
        return ret
    return None


def validate_fb2(string: str):
    """fb2 filename validation"""
    ret = string
    if fb2_check.match(string):
        return ret
    return None


def validate_genre(string: str):
    """genre id validation"""
    ret = string
    if genre_check.match(string):
        return ret
    return None


def validate_search(string: str):
    """search pattern some normalization"""
    if string is None:
        return ""
    ret = unurl(string).replace(';', '')
    if len(ret) > 128:
        ret = ret[:128]
    return ret

File: app_go/util/validate.go
// Package util provides input validation utilities.
package util

import (
	"path/filepath"
	"regexp"
	"strings"
)

// Regex patterns for validation.
var (
	idCheck    = regexp.MustCompile(`^[0-9a-f]+$`)
	zipCheck   = regexp.MustCompile(`^[0-9a-zA-Z_.\-]+\.zip$`)
	fb2Check   = regexp.MustCompile(`^[ 0-9a-zA-ZА-Яа-я_,.:!\-]+\.fb2$`)
	genreCheck = regexp.MustCompile(`^[0-9a-z_]+$`)
)

// ValidateID validates an author/book/sequence ID (hex string).
func ValidateID(s string) string {
	if idCheck.MatchString(s) {
		return s
	}
	return ""
}

// ValidateZip validates a zip filename.
func ValidateZip(s string) string {
	if zipCheck.MatchString(s) {
		return s
	}
	return ""
}

// ValidateFB2 validates an fb2 filename.
func ValidateFB2(s string) string {
	if fb2Check.MatchString(s) {
		return s
	}
	return ""
}

// ValidateGenre validates a genre ID.
func ValidateGenre(s string) string {
	if genreCheck.MatchString(s) {
		return s
	}
	return ""
}

// unurl performs simple URL decoding for specific encoded characters.
func unurl(s string) string {
	if s == "" {
		return ""
	}
	translate := map[string]string{
		"%22": "\"",
		"%27": "'",
		"%2E": ".",
		"%2F": "/",
	}
	ret := s
	for k, v := range translate {
		ret = strings.ReplaceAll(ret, k, v)
	}
	return ret
}

// ValidateSearch normalizes a search pattern.
func ValidateSearch(s string) string {
	if s == "" {
		return ""
	}
	ret := unurl(s)
	ret = strings.ReplaceAll(ret, ";", "")
	if len(ret) > 128 {
		ret = ret[:128]
	}
	return ret
}

// SafePath creates a safe relative path from input (prevents path traversal).
func SafePath(fspath string) string {
	if fspath == "" {
		return ""
	}
	// Join with root and make relative - this prevents ../ traversal
	cleaned := filepath.Clean("/" + fspath)
	// Remove leading "/"
	return cleaned[1:]
}

// ValidatePrefix validates a prefix for authorsindex/sequencesindex URLs.
func ValidatePrefix(s string) string {
	if s == "" {
		return ""
	}
	// Reject path traversal attempts
	ret := SafePath(s)
	if ret != s {
		return ""
	}
	if len(ret) < 1 || len(ret) > 10 {
		return ""
	}
	return ret
}
